File: Natrium/logger.rb
module Esites
  class Logger
    @@show_timestamp = true

    def self._log(line, color=39, should_print=true)
      t = ""
      if @@show_timestamp
        t = "\e[90m[#{Time.now.strftime("%H:%M:%S")}]: ▸\e[0m "
      end
      p = "#{t}\e[#{color}m#{line}\e[0m"
      if should_print == true
        print("#{p}\n")
      end
      return p
    end

    def self.error(line, should_print=true, show_time)
      return _log("[ERROR] #{line}", 31, should_print)
    end

    def self.success(line, should_print=true)
      return _log(line, 92, should_print)
    end

    def self.info(line, should_print=true)
      return _log(line, 93, should_print)
    end

    def self.debug(line, should_print=true)
      return _log(line, 36, should_print)
    end

    def self.log(line, should_print=true)
      return _log(line, 39, should_print)
    end

    private_class_method :_log
  end
end

File: lib/Sources/Logger.swift
//
//  Logger.swift
//  Natrium
//
//  Created by Bas van Kuijck on 07/06/2017.
//
//

import Foundation

class Logger {
    
    static var shouldPrint = true
    
    static var insets: Int = 0
    
    static var logLines: [String] = []
    
    fileprivate static var _dateFormatter: DateFormatter = {
        let dateFormatter = DateFormatter()
        dateFormatter.timeStyle = .medium
        return dateFormatter
    }()
    
    @discardableResult
    fileprivate static func _log(_ line: String,
                                 color: String = "39") -> String {
        let dateString = _dateFormatter.string(from: Date())
        let line = colorWrap(text: "[\(dateString)]: ▸ ", in: "90") +
            String(repeating: "  ", count: insets) +
            colorWrap(text: line, in: color)
        
        if shouldPrint {
            print(line)
        } else {
            logLines.append(line)
        }
        return line
    }
    
    static func colorWrap(text: String, `in` color: String) -> String {
        return "\u{001B}[0;\(color)m\(text)\u{001B}[0m"
    }
    
    @discardableResult
    static func error(_ line: String) -> String {
        return _log("❌  \(line)", color: "31")
    }
    
    @discardableResult
    static func fatalError(_ line: String) -> String {
        insets = 0
        if !shouldPrint {
            let currentDirectory = FileManager.default.currentDirectoryPath
            let filePath = "\(currentDirectory)/Config.swift"
            let contents = "### NATRIUM ERROR ###\n\n\"\(line)\"\n\nSolve this and rebuild"
            FileHelper.write(filePath: filePath, contents: contents)
        }
        _log("❌  \(line)", color: "31")
        exit(EX_USAGE)
    }
    
    @discardableResult
    static func success(_ line: String) -> String {
        return _log("✅  \(line)", color: "92")
    }
    
    @discardableResult
    static func info(_ line: String) -> String {
        return _log(line, color: "93")
    }
    
    @discardableResult
    static func warning(_ line: String) -> String {
        return _log("⚠️  \(line)", color: "40;38;5;215")
    }
    
    @discardableResult
    static func debug(_ line: String) -> String {
        return _log(line, color: "36")
    }
    
    @discardableResult
    static func log(_ line: String) -> String {
        return _log(line, color: "39")
    }
    
    @discardableResult
    static func verbose(_ line: String) -> String {
        return _log(line, color: "37")
    }
}
